### DISCUSSION TITLE
No Audio using when using go2rtc

### DISCUSSION THREAD

**Discussion by noloboy** [gtxaspec/wz_mini_hacks#664] [Q&A]
Hi!  

I successfully install wz_minihack on my Wyze Cam v3 and was also able to use go2rtc links like the mp4 link (for HASS or Frigate), Webrtc webstream, MSE, etc.

For some reason, there is no audio coming out from any of the go2rtc links.  Video is perfectly fine.

Using the RTSP from WZ_minihack, I get audio.

Any suggestions for me to hear audio coming from the camera using any of the go2rtc links?

TIA!

**Comment by shamlord** [comment]
I'm seeing the same issue. Go2RTC RTSP streams have no audio from the web links on the camera, or in Blue Iris.

**Comment by gtxaspec** [comment]
which audio format do you have enabled in go2rtc?

**Reply by shamlord** [reply]
I'm using the default go2rtc config:
```
log:
  level: info # default level
  api: trace
  exec: debug
  ngrok: info
  rtsp: warn
  streams: error
  webrtc: fatal
ffmpeg:
  bin: /opt/wz_mini/bin/ffmpeg
streams:
  1080p:
# NOTE: USE /dev/video6 for Wyze Cam v2/PanV1!
  - exec:ffmpeg -fflags +genpts+flush_packets+nobuffer+bitexact -flags +low_delay -thread_queue_size 100 -analyzeduration 50 -probesize 50 -pix_fmt yuv420p -r 30 -i /dev/video1  -flags +global_header -movflags +empty_moov+default_base_moof+frag_keyframe -framerate 30 -video_size 1920x1080 -pix_fmt yuv420p -c:v copy -copyts -rtsp_transport tcp -f rtsp {output}
  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper0 -c:a copy -muxdelay 0 -muxpreload 0 -f rtsp {output}
  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper0 -c:a libfdk_aac -afterburner 1 -channels 1 -b:a 128k -profile:a aac_he -ar 16000 -strict experimental -muxdelay 0 -muxpreload 0 -ac:a 1 -f rtsp {output}
# libopus has higher cpu requirements, audio may studder.
#  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper0 -c:a libopus -muxdelay 0 -muxpreload 0 -ac:a 1 -af adelay=0|0 -f rtsp {output}
  360p:
# NOTE: USE /dev/video7 for Wyze Cam v2/PanV1!
  - exec:ffmpeg -fflags +genpts+flush_packets+nobuffer+bitexact -flags +low_delay -thread_queue_size 100 -analyzeduration 50 -probesize 50 -pix_fmt yuv420p -r 30 -i /dev/video2 -framerate 30 -video_size 640x360 -pix_fmt yuv420p -c:v copy -copyts -rtsp_transport tcp -f rtsp {output}
  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper2 -c:a copy -muxdelay 0 -muxpreload 0 -ac:a 1 -f rtsp {output}
  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper2 -c:a libfdk_aac -afterburner 1 -channels 1 -b:a 128k -profile:a aac_he -ar 16000 -strict experimental -muxdelay 0 -muxpreload 0 -ac:a 1 -f rtsp {output}
# libopus has higher cpu requirements, audio may studder.
#  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper2 -c:a libopus -muxdelay 0 -muxpreload 0 -ac:a 1 -af adelay=0|0 -f rtsp {output}
```

I'm only using the 1080p stream, so it appears that the codecs are either "copy" or "libfdk_aac". I don't understand why there are three ffmpeg processes in the config file. If I watch a live stream from the camera through BlueIris, I hear no audio even with live audio enabled for that camera. If I trigger a recording, the video clip has static for audio. It's not clipping like the volume or gain are too high either, its just garbage audio. The video clip is normal however.

**Comment by gtxaspec** [comment]
You can provide multiple codecs to the streams, and if you are using the web player, it will choose the appropriate codec.     If you are using WebRTC, then you should use OPUS or PCM  @48000hz.   if using MSE, you should be able to use AAC, PCM, OPUS.  

You don't have to have multiple codecs served, you can remove lines from the config.   Checkout the go2rtc docs at https://github.com/AlexxIT/go2rtc for more detailed information.

**Reply by shamlord** [reply]
My only use case for go2rtc would be to serve an RTSP stream with audio to Blue Iris. I'm not terribly interested in using the web links on the camera. I did actually get audio working in VNC with the following link:  ```rtsp://172.16.0.32:8554/1080p?mp4``` but this stream did not work in Blue Iris. Still usable video but not audio. I'm not trying to do anything other than use go2rtc as a drop in replacement for the previous rtsp server.  I entirely expected to set go2rtc_high_res=true and high_res_audio=true, copy the RTSP url into Blue Iris, and call it a day. Instead I've spent at least 6 hours of spinning my wheels and googling on this one simple problem of not having audio recorded from my cameras.

The MP4 file output from the camera uses PCM ALAW, mono audio, at 16,000hz and 16 bits. I've tried both a-Law and u-Law codecs on the camera configuration in BI. I've also tried the "RTSP" and "RTSP alternate AAC decoder" audio settings in the BI camera config dialog. It never changes the resulting audio.  Commenting out one of the go2rtc config lines which appeared to be solely related to audio actually broke the video stream as well in my testing with VLC. Trying to set ```pcma``` or ```pcmu``` as audio codecs in the YAML file also get me nowhere. Trying to make heads or tails of the default config using the go2rtc project's documentation is another dead end. The project's config file examples look nothing like the configuration included with the current version of wz_mini. Their documentation goes into the minutiae of each individual option without explaining the overall structure of the config file, I would think that each individual ffmpeg line configures an individual stream, but commenting out any of the 1080p ffmpeg lines breaks the video stream as well. go2rtc may be an excellent replacement for the old RTSP server, but its config file is just nuts. I was surprised to find one of the ffmpeg commands is over 350 characters long.

According to the go2rtc documentation, the RTSP client (Blue Iris) should negotiate the stream with the server. That's certainly not happening. I am tempted to just empty the go2rtc config file and try creating my own, but I am stopped by my conclusion that since wz_mini is released specifically for Wyze cameras, I presume the included go2rtc configuration is somehow optimized for their output, is it not?

**Reply by shamlord** [reply]
After having some sleep, and also reading some comments on forum posts related to using go2rtc with Frigate, and then sitting down with the ffmpeg documentation and going through each command line switch, **I was able to get audio to work in Blue Iris**.

I found the following bits of info that non developers (like myself) may find useful.

In go2rtc parlance, "streams" are what go2rtc receives audio and video **_from_**. "Modules" are what go2rtc sends video and audio **_to_**. This is a bit confusing when you read the go2rtc documentation, because of the versatility of go2rtc. It is possible to receive data from an RTSP _stream_ and send it to the RTSP _module_. Something else is that most if not all the modules are active by default. So you don't actually pipe the audio and video to a given module, rather you just connect to the port go2rtc is listening to for a given protocol and go2rtc figures it out. 

If I understand correctly, there are three simultaneous ffmpeg instances in wz_mini's default config for 1080p. The video input stream is received by running ffmpeg on the output of ```/dev/video1```.  The second and third ffmpeg instances run on the output of ALSA; with the 2nd line direct copying the audio, and the third line transcoding to AAC.

For the RTSP module, go2rtc will serve the first audio/video stream that may be compatible with the connected client, starting at the top of the config file. So in my case, it go2rtc would use the first audio ffmpeg line which uses the ```-c:a copy``` switch to just send the Wyzecam's raw audio a-Law output to the output modules. This doesn't work with Blue Iris unfortunately. VLC does not even show an audio stream when using ```-c:a copy```, so is RTSP even compatible with that audio format? I don't know. There may be a way to adjust the original audio stream from the camera to work, but for now I settled on transcoding audio to AAC:

```
log:
  level: info # default level
  api: trace
  exec: debug
  ngrok: info
  rtsp: warn
  streams: error
  webrtc: fatal
ffmpeg:
  bin: /opt/wz_mini/bin/ffmpeg
streams:
  1080p:
  - exec:ffmpeg -fflags +genpts+flush_packets+nobuffer+bitexact -flags +low_delay -thread_queue_size 100 -analyzeduration 50 -probesize 50 -pix_fmt yuv420p -r 30 -i /dev/video1  -flags +global_header -movflags +empty_moov+default_base_moof+frag_keyframe -framerate 30 -video_size 1920x1080 -pix_fmt yuv420p -c:v copy -copyts -rtsp_transport tcp -f rtsp {output}
#  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper0 -c:a copy -muxdelay 0 -muxpreload 0 -f rtsp {output}
  - exec:ffmpeg -hide_banner -f alsa -ac 1 -i dsnooper0 -c:a libfdk_aac -afterburner 1 -channels 1 -b:a 128k -profile:a aac_he -ar 16000 -strict experimental -muxdelay 0 -muxpreload 0 -ac:a 1 -f rtsp {output}
```
The main changes are that I commented out the direct audio copy ffmpeg line, and I removed the 360p video stream as I am in a very congested 2.4ghz environment. I have the CPU to use the 1080p stream for object detection but not the bandwidth for a substream.

For clarity, I am using Blue Iris 5.8.1.4.  They do not offer a detailed change log, but it seems support for AAC audio in RTSP streams is a recent change around 5.7.x.x. Attached is the camera config dialog from BI.

![Cam Config](https://github.com/gtxaspec/wz_mini_hacks/assets/75527296/b3e518e4-feeb-421e-b847-2b99951c2845)
